#116: Aggregate Variation (DiscreteVariationSet, Haplotype, Genotype)

reece opened the pull request:
This is a *draft* pull request. That means that I'm soliciting feedback, but it's not ready to merge.

See https://vr-spec.readthedocs.io/en/116-variation-sets/terms_and_model.html for formatted docs

To do list from comments below:
- [x] flatten hierarchy / drop AggregateVariation and DiscreteVariation (@larrybabb, @ahwagner, @reece )
- [x] improve variation id description (@melissacline )
- [x] clarify use of CURIE in VariationSet (@andrewyatz )
- [x] add prefix to computed identifier discussion (@larrybabb )
- [x] update schema diagram (@rrfreimuth )
- [x] investigate whether links to vr.yaml and vr.json can be relative to current branch (@andrewyatz )

larrybabb commented:
Thank you for getting this work out for review @reece it is greatly appreciated.

I'm not comfortable with the leap from Alleles and Haplotypes (which seem to be one and the same) to Genotypes. 

Alleles and Haplotypes are an explicit representation of zero or more state(s) on a single instance of a molecular sequence.
Genotypes are a representation of zero or more state(s) for the same regions across multiple instances of a molecular sequence within the context of a system (i.e. genome, cell, sample, etc...).

I believe that we can and should reduce complexity in the model by representing Alleles and Haplotypes with a single structure.

I also believe that we should separate the variation of change on a single instance of a molecular sequence from changes across all occurrences within a system (i.e. genome) so that we clarify that Haplotypes/Alleles are components of more complex variation.

Mixing genotypes and haplotypes/alleles under a single grand concept of variation is useful in the big picture, but it is confusing for use cases in defining observed variation in somatic and germline studies - I think. 

Folks will want to place knowledge on either but need to clearly know when they are making a Allele/Haplotype reference that requires additional effort when applying to observed Genotype findings and vice versa.

The other value of segregating Genotype from Allele/Haplotype is that it will provide the opportunity to address other kinds of variation notions that are constrained by the context of a given "system" (i.e. genome, tissue, cell,...). For example, CNVs.

Structural Variation may be yet a third (separate) form of variation that we can define in the context of several Allele/Haplotype changes (i.e. translocations, duplications, deletions), the key point here being that we are not expressing exact sequences observed, merely regional impacts.  

While it is true that a structural variant can also be a sequence variant, it doesn't mean that we should use only one of those concepts to represent both forms of variation. By separating these I believe we will provide clarity and direction to the share representations of these fundamental concepts of molecular variation that the field is doing in a variety of confusing and computationally insufficient ways today.

larrybabb commented:
Can we say that genotypes are a higher-order form of variation within the context of a system that is comprised only of exact sequence variation (haplotypes/alleles)?  I don't think we want to include structural variation in the lower level components that define a genotype. This would be great to verify with you both.

reece commented:
@larrybabb: Please draw the hierarchy that you envision so that we have a concrete proposal to discuss.

ahwagner commented:
Thanks for this, @reece. We have discussed prior, but I am not comfortable / clear on the use of DiscreteVariaton and AggregateVariation as formal classes in the specification. I think I was unavailable for the conversation around this due to travel (and I am always reluctant to push back against substantial effort) but I suggest that we drop the notions of DiscreteVariation and AggregateVariation as classes in the schema.

From the look of it, there are no technical attributes that are inherited from the abstract [DiscreteVariation](https://github.com/ga4gh/vr-spec/blob/116-variation-sets/schema/vr.yaml#L129-L138) and [AggregateVariation](https://github.com/ga4gh/vr-spec/blob/116-variation-sets/schema/vr.yaml#L205-L209) classes, only notions that certain forms of Variation belong to one of these classes. However, by introducing these we have now constrained our [Variation superclass](https://github.com/ga4gh/vr-spec/blob/116-variation-sets/schema/vr.yaml#L113-L122) to:
> A polymorphic representation of Discrete Variation and Aggregate Variation

As evidenced by our discussions this week (including some comments from @larrybabb above), there are multiple ways that people might want to group the Variation subclasses (e.g. molecular, systemic, technical), and these groupings may increase as we develop more Variation subclasses. @reece, I believe you mentioned that there is no clear way to support multiple such groupings in JSON Schema. If I understood that correctly, then I believe we should not restrict our schema to any particular concept grouping unless there is a critical need for it.

reece commented:
Per discussion w/LB and AW: 
1. `Haplotype` and `Genotype` models were moved to a separate branch (`186-haplotype-genotype`).
1. The hierarchy was flattened to remove `DiscreteVariation` and `AggregateVariation`. `Text`, `Allele`, and `VariationSet` are all immediate subclasses of `Variation`.

reece commented:
P.S. https://vr-spec.readthedocs.io/en/116-variation-sets/terms_and_model.html is up-to-date.

reece commented:
@rrfreimuth noticed that the schema diagram needs to be updated. @reece

reece commented:
@andrewyatz : links in machine readable specs point to HEAD, not current branch. Possible to fix?

reece commented:
Add implementation guidance for CURIEs

Add: prefix for VariationSet in docs (see ga4gh.yaml)

melissacline commented:
I'm confused about the Variation ID in the [Variant Set](https://vr-spec.readthedocs.io/en/116-variation-sets/terms_and_model.html#variation-set) definition.  Does that ID indicate a representative variant, or the variant set?  If it indicates a representative variant, the best practices for choosing the representative feels like a can of worms itself, although one that could be tackled.  If the ID is for the variant set, an example would greatly help, and the description should be updated.  Thanks!

helensch commented on docs/source/terms_and_model.rst:
```diff
@@ -698,3 +703,60 @@ subclasses, but are still treated as variation.
 .. _genetic variation: https://en.wikipedia.org/wiki/Genetic_variation
+
+
+Variation Set
```
Should this be VariationSet (without spaces, so that same as SequenceLocation headers

reece replied:
Yep. Another good catch.

reece commented on schema/ga4gh.json:
```diff
@@ -6,6 +6,9 @@
       "regexp": "^ga4gh:(?P<type>[^.]+)\\.(?P<digest>.+)$",
       "type_prefix_map": {
          "Allele": "VA",
+         "DiscreteVariationSet": "DVS",
```
Great catch! Yes, this was *supposed* to have been regenerated with the Makefile. I'm not sure why it wasn't updated. My best guess is that a series of merges and work across concurrent branches left that file newer than the source. I'm a bit bothered by the notion that that's even possible. Fixed.

reece commented:
Thanks for the thorough reading, @helensch! I addressed all of your comments below. The repo and RTD sites are updated.


> * Does the future plans for VariationSets need to be updated?

Yes. Done.

> * The schema diagram https://vr-spec.readthedocs.io/en/116-variation-sets/schema.html#, has asterisk symbols on Allele and VariationSet objects. Does this need to be explained in text below?

Yes. Done.

> * An example of a variation set would be useful as given for other objects in
>   https://vr-spec.readthedocs.io/en/116-variation-sets/terms_and_model.html
> and maybe an example in the implementation guide at same level as hgvs
> https://vr-spec.readthedocs.io/en/116-variation-sets/impl-guide/example.html?highlight=hgvs

Agreed. This was an omission. I added two examples in the terminology page. I chose to not put one in the examples for now, mostly because that's supposed to be a short read to allow people to get a quick take. I worry about adding too much complexity there.

> * A query on nested VariationSets. As VariationSet.members are a list of variation objects, they can be Allele,Text,VariationSet. So nested VariationSets, do not require the members to all be VariationSets?  Thinking of how to represent
>   
>   * a set of all variants say from a study  (variationset VS_A)
>   * a set of all variants from that study  (variationset VS_B) that meet a particular requirement
>     VS_A.members[allele1,allele2 etc, VS_B]
>     or
>     VS_A.members[VS_C,VS_B]
>     VS_C all the variants in the study that do not meet the particular requirement.

The spec is silent about set logic or operations on sets. In the current spec, sets are only additive, so you'd have to build from the bottom up.

I think there are three likely use cases for VariationSet:
* When annotating a group of variants, put them in a set and compute the identifier. Associate the annotation with that identifier.
* When tracking the evolution of a named set, such as a dbSNP record, use a series of variation sets and release dates that describe this evolution.
* In a REST interface, use VariationSet to send or receive a group of variants in a standardized way.

helensch commented:
@reece thanks for all the updates and answering my queries. VariationSets have been removed from https://vr-spec.readthedocs.io/en/116-variation-sets/appendices/future_plans.html. The future-schema diagram is wrt to version 1.0 and VariationSet is shown as future and abstract.  Should the future-schema diagram be updated - or will be fixed and only updated per release?

reece commented:
@helensch Another good catch. I did a minimal fix in 4cc9c5f, but this diagram could probably use a more significant rework to ease congestion.

helensch commented on docs/source/terms_and_model.rst:
```diff
@@ -700,3 +705,145 @@ subclasses, but are still treated as variation.
+  However, objects that use non-`ga4gh` identifiers MAY NOT use the
+  :ref:`computed-identifiers` mechanism.
+* VariationSet identifiers computed using the GA4GH
+  `:ref:`computed-identifiers` process do *not* depend on whether the
```
Maybe an extra ` here as link not displaying on https://vr-spec.readthedocs.io/en/116-variation-sets/terms_and_model.html#variationset

reece replied:
Done. Thanks!

ahwagner commented on docs/source/terms_and_model.rst:
```diff
@@ -700,3 +705,145 @@ subclasses, but are still treated as variation.
+  :ref:`computed-identifiers` process do *not* depend on whether the
+  Variation objects are inlined or referenced, and do *not* depened on
+  the order of members.
+* Do not use VariationSet for Haplotypes or Genotypes. VRS will soon
+  (2020 Q2?) support dedicated classes that will provide appropriate
+  constraints for these data.
```
The current PR already assumes addition of Haplotype and Genotype in `1.1` (see my above comment on the `future_plans.rst` changes). We should omit this to match.
```suggestion
```

reece commented on docs/source/appendices/future_plans.rst:
```diff
@@ -256,251 +255,6 @@ Additional :ref:`Variation` concepts that are being planned for future
 consideration in the specification. See :ref:`Variation` for more
 information.
 
-Haplotypes
```
You're right. Let's leave as-is since H & G seem likely. We'll have to remember to bring this back if we miss including in 1.1.

reece commented on docs/source/impl-guide/computed_identifiers.rst:
```diff
@@ -236,6 +236,7 @@ Type prefixes used by VR are:
 
    SQ, Sequence
    VA, Allele
+   VS, VariationSet
```
That's what I'd like in the future. However, TASC hasn't taken up this charge yet. So, for now, I recommend that we keep this is a VRS thing. Otherwise, they're going to have to create a mechanism for this review (and subsequent ones).

ahwagner replied:
I agree, but want GKS leads to weigh in too. Started up a conversation in Slack on #gks_leads.

reece commented on docs/source/rst_epilog:
```diff
@@ -21,8 +21,8 @@
-.. |vr_json| replace:: `vr.json <https://raw.githubusercontent.com/ga4gh/vr-spec/{release}/schema/vr.json>`__
-.. |vr_yaml| replace:: `vr.yaml <https://raw.githubusercontent.com/ga4gh/vr-spec/{release}/schema/vr.yaml>`__
+.. |vr_json| replace:: :download:`vr.json <_static/vr.json>`
+.. |vr_yaml| replace:: :download:`vr.yaml <_static/vr.yaml>`
```
That's quite a statement for RST, which doesn't lend itself to elegent often! (And thanks)

reece commented on docs/source/terms_and_model.rst:
```diff
@@ -1,5 +1,5 @@
 Terminology & Information Model
-!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Overrun isn't significant. It's probably a vestige of an edit.

Although, given that it's RST, you might expect that it causes wild behavior, like translation to Arabic or Urdu.
